### PR TITLE
add: nested filter input-types

### PR DESCRIPTION
**Pull request by hedwiggggg**
Hello, in this pull request I added the possibility to create filter types at any depth. So far I have only run the unit-tests.

So still missing:
- [x] run e2e-tests
- [x] add unit-tests
- [x] add e2e-tests
- [x] add documentation

Internal tests already ran successfully, but I noticed a bug with the query builder. In the generated joins the alias is set to the filter-property. So a hypothetical filter like this:

```typescript
const articlesFilter: ArticleFilter = {
  group: {
    articles: {
      group: {
        name: {
          eq: 'tools',
        },
      },
    },
  },
};
```

Ends with the error: `table name "group" specified more than once`. Here is the log:

```
query failed: SELECT "Article"."id" AS "Article_id", "Article"."created_at" AS "Article_created_at", "Article"."updated_at" AS "Article_updated_at", "Article"."deleted_at" AS "Article_deleted_at", "Article"."status" AS "Article_status", "Article"."order_no" AS "Article_order_no", "Article"."name" AS "Article_name", "Article"."description" AS "Article_description", "Article"."condition" AS "Article_condition", "Article"."weight" AS "Article_weight", "Article"."unit" AS "Article_unit", "Article"."is_sellable" AS "Article_is_sellable", "Article"."is_storable" AS "Article_is_storable", "Article"."is_discountable" AS "Article_is_discountable", "Article"."group_id" AS "Article_group_id", "Article"."country_of_origin" AS "Article_country_of_origin", "Article"."customs_tariff_number" AS "Article_customs_tariff_number", "Article"."supplier_id" AS "Article_supplier_id", "Article"."supplier_order_number" AS "Article_supplier_order_number", "Article"."designation" AS "Article_designation" FROM "article" "Article" LEFT JOIN "article_group" "group" ON "group"."id"="Article"."group_id" AND ("group"."deleted_at" IS NULL)  LEFT JOIN "article" "articles" ON "articles"."group_id"="group"."id" AND ("articles"."deleted_at" IS NULL)  LEFT JOIN "article_group" "group" ON "group"."id"="articles"."group_id" AND ("group"."deleted_at" IS NULL) WHERE ( (((((("group"."name" = $1))))) AND 1=1 AND 1=1) ) AND ( "Article"."deleted_at" IS NULL ) ORDER BY "Article"."id" ASC LIMIT 26 -- PARAMETERS: ["Tools"]
```

But this is a problem for an own issue / pull request.

### CODE REVIEW

**Comment by hedwiggggg** [comment]
To configure the filter-depth for an entity, you pass the `filterDepth` property to the `QueryOptions`-decorator.

```typescript
@ObjectType('TodoItem')
// `filterDepth` is a number, 1 is the default
@QueryOptions({ filterDepth: 3 })
// `filterDepth: Number.POSITIVE_INFINITY` will result in a infinitely deep filter
@QueryOptions({ filterDepth: Number.POSITIVE_INFINITY })
export class TodoItemDTO {
  @IDField(() => ID)
  id!: number;

  @FilterableField()
  title!: string;

  @FilterableField({ nullable: true })
  description?: string;

  @FilterableField()
  completed!: boolean;

  @FilterableField(() => GraphQLISODateTime)
  created!: Date;

  @FilterableField(() => GraphQLISODateTime)
  updated!: Date;
}
```

**Comment by TriPSs** [comment]
Thanks for the PR! Could you show me a query filter example that before did work and does not now? I'm not quite sure I completely understand what this does.

**Comment by hedwiggggg** [comment]
This is how the generated schema will look like if the `filterDepth` is `Number.POSITIVE_INFINITY`.

image

**Comment by hedwiggggg** [comment]
This is how the generated schema will look like if the `filterDepth` is `2`.

image

File: packages/query-graphql/src/types/query/filter.type.ts [hunk 1]
```diff
@@ -12,30 +12,48 @@ import { createFilterComparisonType } from './field-comparison'
...
+  Object.keys(relations).forEach((r) => {
+    const opts = relations[r]
+    if (opts && opts.allowFiltering) {
+      filterableRelations[r] = opts.DTO
+    }
+  })
+  return filterableRelations
+}
+
+function getOrCreateFilterType<T>(TClass: Class<T>, name: string, depth = 0): FilterConstructor<T> {
```
Comment (TriPSs) [inline packages/query-graphql/src/types/query/filter.type.ts:47]: Since we call `getOrCreateFilterType` already with dept = 1 do we need the `dept = 0` here?

Comment (hedwiggggg) [reply packages/query-graphql/src/types/query/filter.type.ts:47]: Yes, because we call it with `depth = 1` only on the `(Delete / Update / Subscription / Aggregate)FilterType`. The base `FilterType` is called without an argument.

What we can do is to remove the default value and move the `depth = 0` into the base `FilterType`.

**Comment by TriPSs** [comment]
So if I understand correctly, default it create a `AddressFilterBankFilter` so you can filter the address on the bank relation, if `Number.POSITIVE_INFINITY` is used it will use the "normal" `BankFilter` with all the filters that `Bank` has. And if you provide `filterDepth = 2` it will create a other one so you can filter `address -> bank(1) -> other bank relation(2)` correct?

**Comment by hedwiggggg** [comment]
> So if I understand correctly, default it create a `AddressFilterBankFilter` so you can filter the address on the bank relation, if `Number.POSITIVE_INFINITY` is used it will use the "normal" `BankFilter` with all the filters that `Bank` has. And if you provide `filterDepth = 2` it will create a other one so you can filter `address -> bank(1) -> other bank relation(2)` correct?

Exactly.

**Comment by TriPSs** [comment]
And a question: If we where to release this like this it will not change any schema except if you provide the `filterDepth`?

File: packages/query-graphql/src/types/query/filter.type.ts (hunk 1 above)
Comment (TriPSs) [reply packages/query-graphql/src/types/query/filter.type.ts:47]: Okay check, maybe indeed call it there with `dept = 0` so we don't need the default.

**Comment by hedwiggggg** [comment]
> And a question: If we where to release this like this it will not change any schema except if you provide the `filterDepth`?

Nope, otherwise the test wouldn't have passed.

**Comment by hedwiggggg** [comment]
Should be ready now @TriPSs 🚀

**Comment by TriPSs** [comment]
Thanks for the PR! Changes look good, I do want to check this out locally before merging if you don't mind.

**Comment by TriPSs** [comment]
@hedwiggggg after checking this locally I did notice one thing:
```ts
@FilterableRelation('relation', () => Relation, {
  // This one does not work
  filterDepth: 2,
})
```

Can we either remove it or add support for it? (I do think it's nice we can specific it for a specific relation instead of all relations)

**Comment by hedwiggggg** [comment]
Unfortunately, this can only be realized with breaking changes, since the target depth would have to be embedded in the filter name. Otherwise, it cannot be guaranteed which filter depth will be generated because the root filter names are the same for different depths. And configuring each relation individually, would possibly lead to different depths for the same filter type.

I have specially designed this feature so that there are no breaking changes.

**Comment by hedwiggggg** [comment]
tldr; we would need to generate `n` filter-types for the same entity where `n` is equal to the count of the unique `{Entity, filterDepth}` occurrences. 

In the following example we would need 3 different filter types, but in the current implementation all filter types would have the same name, which is invalid in GraphQL.

```
// Example-A.entity.ts
@FilterableRelation('relation', () => Relation, { filterDepth: 2 })
```

```
// Example-B.entity.ts
@FilterableRelation('relation', () => Relation)
```

```
// Example-C.entity.ts
@FilterableRelation('relation', () => Relation, { filterDepth: 3 })
```

**Comment by TriPSs** [comment]
Understandable, could we then remove it from there so the option is not shown?

**Comment by hedwiggggg** [comment]
I thought about it again last night. There is a way to configure this per relation, without breaking-changes. By default the filter depth is 1. With this filter depth the generated name can also be `${ObjectTypeName}Filter` by default. If the depth differs from the default filter depth, then we embed the target depth in the name.

```
// Example-A.entity.ts
@ObjectType('EntityA')
@FilterableRelation('relation', () => relation)
class EntityA { /* ... */ }

// leads to
// input EntityAFilter {
// ...
// }
```

```
// Example-A.entity.ts
@ObjectType('EntityA')
@FilterableRelation('relation', () => relation, { filterDepth: 2 )
class EntityA { /* ... */ }

// leads to
// input EntityADepth2Filter {
// ...
// }
```

But nevertheless I would take it out for now and put the feature-request into a separate issue, since I don't have the resources for it right now.

**Comment by TriPSs** [comment]
@hedwiggggg massive thanks for this addition!
